src/Workspaces/Core/Portable/Workspace/Host/TemporaryStorage/ITemporaryStorageService.cs

```diff
@@ -16,6 +18,40 @@ public interface ITemporaryStorageService : IWorkspaceService
 
 internal interface ITemporaryStorageServiceInternal : IWorkspaceService
 {
-    ITemporaryStreamStorageInternal CreateTemporaryStreamStorage();
+    TemporaryStorageHandle WriteToTemporaryStorage(Stream stream, CancellationToken cancellationToken);
+    Stream ReadFromTemporaryStorageService(TemporaryStorageIdentifier storageIdentifier, CancellationToken cancellationToken);
+
     ITemporaryTextStorageInternal CreateTemporaryTextStorage();
 }
```

Comment:
here's the crux of the change.  The prior API (CreateTemporaryStorageStream) would effectively give you an instance of an object you could read or write to.  But that object was super strange.  For example, you couldn't actually read from this object (since nothing had written to it).  And, if you wrote to it, you could only write to it once. 

Furthermore, that value returned happened to implement *other* APIs as well taht would allow passing information through a backdoor to help inform OOP serialization.  Specifically, what the mmf information was for the written data so we could send that data over to OOP to let it hydrate the info on its side.

The API changes heavily now.  First off, when you want to write some stream to a mmf, you just say `WriteToTemporaryStorage` passing in the data you actually want to write *up front*.  This gives you back a 'handle' representing that data.  The handle allows getting the mmf data for cases like OOP.  It also supports releasing that written data when no longer needed.

Interestingly enough though, that 'releasing' behavior is only used in *exactly* one place in roslyn.  And that place is unrelated to SourceTexts or References, and is unrelated to communicating with OOP.  It's literally where we dump the *enormous* compiler-command-line string we build that we don't want actually staying in memory.    All the OOP cases, and all the SourceText/Reference mmf dumping never releases the data from the memory mapped files.  This is likely a leak in the traditional sense.  But one that is so slow that it is rarely noticed in practice (no one has ever mentioned it up till now).  

Because the data is never released, the protocol with the server is simple, as the server can just depend on this data stayign alive. 

--

The next method that is added is the ReadFromTemporaryStorageService method.   This takes the information produced by the WriteToTemporaryStorage, using it to erad from teh MMF and supply the data in a stream.